File: sort/heap_sort.go
package sort

func maxHeapify(inArr []int, startIdx, endIdx int) {
	dadIdx := startIdx
	sonIdx := dadIdx*2 + 1
	for sonIdx <= endIdx {
		if sonIdx+1 <= endIdx && inArr[sonIdx] < inArr[sonIdx+1] {
			sonIdx++
		}
		if inArr[dadIdx] > inArr[sonIdx] {
			return
		}
		inArr[dadIdx], inArr[sonIdx] = inArr[sonIdx], inArr[dadIdx]
		dadIdx = sonIdx
		sonIdx = dadIdx*2 + 1
	}
}

// heap sort - asc
func HeapSort(inArr []int) {
	len := len(inArr)
	for i := int(len/2) - 1; i >= 0; i-- {
		maxHeapify(inArr, i, len-1)
	}
	for i := len - 1; i > 0; i-- {
		inArr[0], inArr[i] = inArr[i], inArr[0]
		maxHeapify(inArr, 0, i-1)
	}
}

File: sort/heap_sort.cpp
// +build ignore
// from golang

#include "heap_sort.h"

using namespace std;

void maxHeapify(int arr[], int startIdx, int endIdx) {
    int dadIdx = startIdx, sonIdx = dadIdx * 2 + 1;
    while (sonIdx <= endIdx) {
        if (sonIdx + 1 <= endIdx && arr[sonIdx] < arr[sonIdx + 1]) sonIdx++;
        if (arr[dadIdx] > arr[sonIdx]) return;
        swap(arr[dadIdx], arr[sonIdx]);
        dadIdx = sonIdx;
        sonIdx = dadIdx * 2 + 1;
    }
}
 
void heapSort(int arr[], int len) {
    for (int i = len / 2 - 1; i >= 0; i--)
        maxHeapify(arr, i, len - 1);
    for (int i = len - 1; i > 0; i--) {
        swap(arr[0], arr[i]);
        maxHeapify(arr, 0, i - 1);
    }
}